### PR TITLE
Add Password Hashing and Standardized Error Handling

### LINKED ISSUE #26: Error Handling
When an error is thrown somewhere in our code, we need a standardized way to handle it. This includes catching it, performing any needed operations, and then sending a response to the client (frontend) about what happened.

### LINKED ISSUE #27: Password Hashing
Hash passwords on user create and validate hash on login.

### PR DESCRIPTION
**Pull request by suneettipirneni**
Overview of changes:

- Responses now omit the `password` field for security reasons.
- All errors are now handled via a custom json handler
     - For example if we run into an error:
     ```js
     try {
        // Something that might throw an error.
     } catch(error) {
       // Serialize error
       const error = new APIError(error.title);
       // Pass to error handler.
       return next(error);
     }
     ```
   - That would then generate a json which looks like:
   ```json
   {
    "status": 500,
    "title": "Error Title",
    "description": "The error description",
    "instance": null
   }
   ```
   - Note: `instance` is only used when the request tries to operate on a specific resource.
- Passwords are now hashing, this means any logins used before this is merged will not be able to login after this merges.

Closes #27, Closes #26

### CODE REVIEW

**Review by Armintron — CHANGES_REQUESTED** [review]
Overall good, just some minor changes.

File: src/controllers/group.js [hunk 1]
```diff
@@ -16,33 +18,45 @@ const Group = {
...
+      );
     }
+
+    // Check if user is authorized to join.
     const user = ObjectId(req.body.user);
     const authorizedUser = (group.users).some(x => x.equals(user));
+
     if (group.creator.equals(user) || authorizedUser) {
-      return res.send({ group, message: 'SUCCESSFULLY AUTHENTICATED' });
+      return res.status(204).send({ group, message: 'SUCCESSFULLY AUTHENTICATED' });
```
[the remark this answers is not in the document]
Comment (suneettipirneni) [reply src/controllers/group.js:51]: Logically there's now difference, but semantically 204 means an existing resource got updated and there is no response body.

[the remark this answers is not in the document]
Comment (Armintron) [reply src/controllers/group.js:51]: But there is a response body. 
```javascript 
send({ group, message: 'SUCCESSFULLY AUTHENTICATED' });
```

File: src/controllers/user.js [hunk 2]
```diff
@@ -1,43 +1,80 @@
...
     // create new user model with given request body
     const newUser = new UserModel(
       {
         firstName: req.body.firstName,
         lastName: req.body.lastName,
         username: req.body.username,
-        password: req.body.password,
+        password: hashedPassword,
       },
     );
```
[the remark this answers is not in the document]
Comment (suneettipirneni) [reply src/controllers/user.js:19]: Nice catch!

**Review by Armintron — APPROVED** [review]

**Pull request merged**